statsd/client.py

```diff
@@ -138,14 +151,70 @@ def _send(self, data):
             # No time for love, Dr. Jones!
             pass
 
+    def pipeline(self):
+        return Pipeline(self)
+
+
+class TCPStatsClient(StatsClientBase):
+    """TCP version of StatsClient."""
+
+    def __init__(self, host='localhost', port=8125, prefix=None, timeout=None):
+        """Create a new client."""
+        self._host = host
+        self._port = port
+        self._timeout = timeout
+        self._prefix = prefix
+        self._tlocal = threading.local()
+        self._tlocal.sock = None
+
+    def _send(self, data):
+        """Send data to statsd."""
+        if not self._tlocal.sock:
+            self._connect()
+        try:
+            self._do_send(data)
+        except socket.error:
```

Comment:
I'd really rather not do this automatically, it should be up to the application how to continue. If it _wasn't_ a closed connection, this would still retry, which could cause a cascading failure as this suddenly started taking huge amounts of time.

We should provide public `close` and `[re]connect` methods (or even `connect`, `close`, and a `reconnect` that just called them).